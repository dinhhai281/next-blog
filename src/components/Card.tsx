import { Text, VStack, StackProps, forwardRef } from '@chakra-ui/react';
import { fold, Option } from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';
import { Variants } from 'framer-motion';
import React, { useMemo } from 'react';
import { MotionBox } from '.';

export interface CardProps extends StackProps {
  tagLabel: string;
  createdDate: string;
  timeLabel: string;
  label: string;
  summary: string;
  onFocusIn: VoidFunction;
  onFocusOut: VoidFunction;
  isFocus: Option<boolean>;
}

export const Card = forwardRef<CardProps, 'div'>(
  ({ onFocusIn, onFocusOut, isFocus, tagLabel, createdDate, timeLabel, label, summary, ...rest }, ref) => {
    const variants: Variants = {
      active: {
        scale: 1.05,
      },
      inactive: {
        opacity: 0.4,
      },
      none: {
        scale: 1,
      }
    }

    const resolveAnimate = useMemo(() => pipe(
      isFocus,
      fold(() => 'none', value => value ? 'active' : 'inactive')
    ), [isFocus])

    return (
      <MotionBox variants={variants} animate={resolveAnimate}>
        <VStack
          p={10}
          bg='gray.800'
          align='flex-start'
          borderRadius={4}
          {...rest}
          ref={ref}
          onMouseEnter={onFocusIn}
          onMouseLeave={onFocusOut}
        >
          <Text color='white' textTransform='uppercase' letterSpacing={2} fontSize='xs'>
            {tagLabel}
          </Text>
          <VStack align='flex-start' spacing={0}>
            <Text color='gray.500' fontSize='xs' textTransform='capitalize'>
              {createdDate}
            </Text>
            <Text color='gray.500' fontSize='xs' textTransform='capitalize'>
              {timeLabel}
            </Text>
          </VStack>
          <Text color='white' fontSize='2xl' lineHeight='32px'>
            {label}
          </Text>
          <Text color='gray.400' fontSize='sm'>
            {summary}
          </Text>
        </VStack>
      </MotionBox>
    );
  },
);
