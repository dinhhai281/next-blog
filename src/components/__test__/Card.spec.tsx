import { render } from '@testing-library/react';
import { none } from 'fp-ts/lib/Option';
import React from 'react';
import { Card } from '../Card';

describe('Card', () => {
  const renderComponent = () =>
    render(
      <Card tagLabel='tagLabel' createdDate='createdDate' label='label' timeLabel='timeLabel' summary='summary' isFocus={none}/>,
    );

  describe('when render', () => {
    it('should match snapshot', () => {
      const { asFragment } = renderComponent();

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
