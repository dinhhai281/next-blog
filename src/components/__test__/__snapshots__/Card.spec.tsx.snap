// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card when render should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-0"
  >
    <div
      class="chakra-stack css-k7qgmw"
    >
      <p
        class="chakra-text css-wq3tlk"
      >
        tagLabel
      </p>
      <div
        class="chakra-stack css-16xai34"
      >
        <p
          class="chakra-text css-1e4oirs"
        >
          createdDate
        </p>
        <p
          class="chakra-text css-1e4oirs"
        >
          timeLabel
        </p>
      </div>
      <p
        class="chakra-text css-1wqkak5"
      >
        label
      </p>
      <p
        class="chakra-text css-1sjruip"
      >
        summary
      </p>
    </div>
  </div>
</DocumentFragment>
`;
